test/object-store/sync/app.cpp

```diff
@@ -4702,3 +4702,100 @@ TEST_CASE("app: app cannot get deallocated during log in", "[sync][app]") {
     mock_transport_worker.mark_complete();
 }
 #endif
+
+TEST_CASE("app: user logs our while profile is fetched", "[sync][app]") {
+    AsyncMockNetworkTransport mock_transport_worker;
+    enum class TestState { unknown, location, login, profile };
+    struct TestStateBundle {
+        void advance_to(TestState new_state)
+        {
+            std::lock_guard<std::mutex> lk(mutex);
+            state = new_state;
+            cond.notify_one();
+        }
+
+        TestState get() const
+        {
+            std::lock_guard<std::mutex> lk(mutex);
+            return state;
+        }
+
+        void wait_for(TestState new_state)
+        {
+            std::unique_lock<std::mutex> lk(mutex);
```

Comment:
Is there a potential deadlock by locking the mutex here? Maybe lock it after the conditional is notified?
NVM - you need to have the mutex locked before the conditional, and it will free the mutex before blocking.

What about checking to see if it's already in that state before blocking on the conditional? (In case the `app->link_user` completes before the `state.wait_for()`)

Reply:
there was no need for wait_for in the test actually, so I cleaned up that bit. And, if it was in that state already, then the first time the predicate was evaluated it would return true and stop waiting.